src/main/java/org/folio/service/claiming/ClaimingService.java

```diff
@@ -0,0 +1,246 @@
+          .filter(Objects::nonNull).filter(piece -> Objects.nonNull(piece.getId()) && Objects.nonNull(piece.getPoLineId()))
+          .map(piece -> Pair.of(piece.getPoLineId(), piece.getId())).distinct()
+          .toList();
+        logger.info("groupPieceIdsByVendorId:: Prepared unique piece-poLine pairs, pairs: {}", uniquePiecePoLinePairs);
+        return collectResultsOnSuccess(createPieceIdByVendorFutures(pieces, uniquePiecePoLinePairs, requestContext))
+          .map(ClaimingService::transformAndGroupPieceIdsByVendorId);
+      });
+  }
+
+  private List<Future<Pair<String, String>>> createPieceIdByVendorFutures(List<Piece> pieces, List<Pair<String, String>> uniquePiecePoLinePairs,
+                                                                          RequestContext requestContext) {
+    var pieceIdByVendorIdFutures = new ArrayList<Future<Pair<String, String>>>();
+    uniquePiecePoLinePairs.forEach(piecePoLinePairs -> {
+      var foundPiece = pieces.stream()
+        .filter(Objects::nonNull).filter(piece -> Objects.nonNull(piece.getId())).filter(piece -> piece.getId().equals(piecePoLinePairs.getRight()))
+        .findFirst().orElseThrow(() -> new NoSuchElementException(String.format(CANNOT_FIND_A_PIECE_BY_ID, piecePoLinePairs.getRight())));
+      var pieceIdByVendorIdFuture = createVendorPiecePair(piecePoLinePairs, foundPiece, requestContext);
+      if (Objects.nonNull(pieceIdByVendorIdFuture)) {
+        pieceIdByVendorIdFutures.add(pieceIdByVendorIdFuture);
+      }
+    });
+    return pieceIdByVendorIdFutures;
+  }
+
+  private Future<Pair<String, String>> createVendorPiecePair(Pair<String, String> piecePoLinePairs,
+                                                             Piece piece, RequestContext requestContext) {
+    if (Objects.nonNull(piece) && !piece.getReceivingStatus().equals(Piece.ReceivingStatus.LATE)) {
+      logger.info("createVendorPiecePair:: Ignoring processing of a piece not in LATE state, piece id: {}", piece.getId());
+      return Future.succeededFuture();
+    }
+    return purchaseOrderLineService.getOrderLineById(piecePoLinePairs.getLeft(), requestContext)
+      .compose(poLine -> purchaseOrderStorageService.getPurchaseOrderById(poLine.getPurchaseOrderId(), requestContext)
+      .compose(purchaseOrder -> organizationService.getVendorById(purchaseOrder.getVendor(), requestContext)))
+      .map(vendor -> {
+        if (Objects.nonNull(vendor) && Boolean.TRUE.equals(vendor.getIsVendor())) {
+          return Pair.of(vendor.getId(), piecePoLinePairs.getRight());
+        }
+        return null;
+      });
+  }
+
+  private static Map<String, List<String>> transformAndGroupPieceIdsByVendorId(List<Pair<String, String>> piecesByVendorList) {
+    return StreamEx.of(piecesByVendorList).distinct().filter(Objects::nonNull)
+      .groupingBy(Pair::getKey, mapping(Pair::getValue, collectingAndThen(toList(), lists -> StreamEx.of(lists).toList())));
```

Comment:
Why not use directly toList()?
```groupingBy(Pair::getKey, mapping(Pair::getValue, toList()))```

Reply:
> Why not use directly toList()? `groupingBy(Pair::getKey, mapping(Pair::getValue, toList()))`

I stole this from checkIn/receiving, will try your approach too